File: revision_analyses/config/set_paths.m
function cfg = set_paths()
%SET_PATHS  Resolve every data and output location used in this repository.
%
%   cfg = SET_PATHS() reads config/config.json (copy config.example.json and
%   edit it) and returns a struct with absolute paths. All scripts obtain
%   their locations from this function; none contain hard-coded paths.
%
%   Fields returned:
%     cfg.data_root      root of the data distribution
%     cfg.eeg            preprocessed EEG (.set/.fdt)
%     cfg.raw_overt      raw BrainVision recordings (.vhdr/.eeg/.vmrk) from
%                        which the overt blocks are epoched
%                        (optional; used only by 10_overt_arm/preprocessing)
%     cfg.source         source-localised covert data (Destrieux, 148 ROIs)
%     cfg.source_overt   source-localised overt data
%     cfg.audio          Hilbert envelopes of the audio recordings
%     cfg.audio_raw      continuous microphone recordings and event markers
%                        (optional; used only by 11_acoustic_monitoring)
%     cfg.bst_protocol   Brainstorm protocol root (contains anat/ and data/)
%     cfg.out            output root for this repository
%     cfg.eeglab         EEGLAB installation
%     cfg.brainstorm     Brainstorm installation
%     cfg.n_workers      parallel workers to request
%
%   cfg.addback(variant) returns the directory of a component add-back
%   dataset, e.g. cfg.addback('lateye').

here = fileparts(mfilename('fullpath'));
cfgFile = fullfile(here, 'config.json');
if ~isfile(cfgFile)
    error('set_paths:noConfig', ...
        'config.json not found. Copy config.example.json to config.json and edit it.');
end

raw = jsondecode(fileread(cfgFile));

root = raw.data_root;
if ~isfolder(root)
    error('set_paths:badRoot', 'data_root does not exist: %s', root);
end

cfg = struct();
cfg.data_root    = root;
cfg.eeg          = fullfile(root, raw.eeg_dir);
if isfield(raw, 'raw_overt_dir'), cfg.raw_overt = fullfile(root, raw.raw_overt_dir); else, cfg.raw_overt = ''; end
cfg.source       = fullfile(root, raw.source_dir);
cfg.source_overt = fullfile(root, raw.source_overt_dir);
cfg.audio        = fullfile(root, raw.audio_dir);
if isfield(raw, 'audio_raw_dir'), cfg.audio_raw = raw.audio_raw_dir; else, cfg.audio_raw = ''; end
cfg.bst_protocol = raw.brainstorm_protocol;
cfg.eeglab       = raw.eeglab_dir;
cfg.brainstorm   = raw.brainstorm_dir;
cfg.n_workers    = raw.n_workers;

outRoot = raw.output_root;
if ~isAbsolutePath(outRoot)
    outRoot = fullfile(fileparts(here), outRoot);
end
cfg.out = outRoot;
if ~isfolder(cfg.out), mkdir(cfg.out); end

pattern = raw.addback_dir_pattern;
cfg.addback = @(variant) fullfile(root, sprintf(pattern, variant));

end


function tf = isAbsolutePath(p)
if ispc
    tf = numel(p) > 1 && (p(2) == ':' || startsWith(p, '\\'));
else
    tf = startsWith(p, '/');
end
end

File: revision_analyses/config/set_paths.py
"""Resolve every data and output location used in this repository.

Copy ``config.example.json`` to ``config.json``, edit it, then::

    from config.set_paths import set_paths
    cfg = set_paths()
    env = cfg.audio / "S0009_envelope.mat"

All scripts obtain their locations from :func:`set_paths`; none contain
hard-coded paths.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path

_HERE = Path(__file__).resolve().parent


@dataclass(frozen=True)
class Config:
    data_root: Path
    eeg: Path
    raw_overt: Path
    source: Path
    source_overt: Path
    audio: Path
    audio_raw: Path
    bst_protocol: Path
    out: Path
    eeglab: Path
    brainstorm: Path
    n_workers: int
    _addback_pattern: str

    def addback(self, variant: str) -> Path:
        """Directory of a component add-back dataset, e.g. ``addback('lateye')``."""
        return self.data_root / (self._addback_pattern % variant)


def set_paths(config_file: Path | None = None) -> Config:
    cfg_path = Path(config_file) if config_file else _HERE / "config.json"
    if not cfg_path.is_file():
        raise FileNotFoundError(
            "config.json not found. Copy config.example.json to config.json and edit it."
        )

    raw = json.loads(cfg_path.read_text())
    root = Path(raw["data_root"]).expanduser()
    if not root.is_dir():
        raise FileNotFoundError(f"data_root does not exist: {root}")

    out = Path(raw["output_root"]).expanduser()
    if not out.is_absolute():
        out = _HERE.parent / out
    out.mkdir(parents=True, exist_ok=True)

    return Config(
        data_root=root,
        eeg=root / raw["eeg_dir"],
        raw_overt=root / raw.get("raw_overt_dir", ""),
        source=root / raw["source_dir"],
        source_overt=root / raw["source_overt_dir"],
        audio=root / raw["audio_dir"],
        audio_raw=Path(raw.get("audio_raw_dir", "")).expanduser(),
        bst_protocol=Path(raw["brainstorm_protocol"]).expanduser(),
        out=out,
        eeglab=Path(raw["eeglab_dir"]).expanduser(),
        brainstorm=Path(raw["brainstorm_dir"]).expanduser(),
        n_workers=int(raw["n_workers"]),
        _addback_pattern=raw["addback_dir_pattern"],
    )
